### ISSUE TITLE
[Security] [Firewall] Still cannot return null from SimplePreAuthenticationListener::createToken() 

### ISSUE THREAD

**Issue by dadamssg** [symfony/symfony#13751]
I've found a couple of issues regarding [only authenticating for certain urls](http://symfony.com/doc/current/cookbook/security/api_key_authentication.html#only-authenticating-for-certain-urls) when using the `SimplePreAuthenticationListener`.

The docs claim you can return `null` from the `::createToken()` method and I've found a [pull request](https://github.com/symfony/symfony/pull/12002) that supposedly fixes this bug. However, the bug persists because the `AccessListener` kicks in an [throws an exception](https://github.com/symfony/Security/blob/master/Http/Firewall/AccessListener.php#L53) if there's no token found. 

Unfortunately I don't know the security component well enough to take a crack at a pull request.

The issues referenced above
[#3815](https://github.com/symfony/symfony-docs/issues/3815)
[#1033](https://github.com/symfony/symfony-docs/issues/1033)

**Comment by Nek-** [comment]
Has you have to return a token or throw an exception. That's quite normal. See this cookbook that looks ok: http://symfony.com/doc/current/cookbook/security/api_key_authentication.html#the-api-key-authenticator

But this issue looks more like a symfony-docs one. Please report this to the [correct repository](https://github.com/symfony/symfony-docs/).

But it looks like there's no codeblock on [the interface](https://github.com/symfony/symfony/blob/2.7/src/Symfony/Component/Security/Core/Authentication/SimpleAuthenticatorInterface.php), and that's not so cool, indeed.

**Comment by dadamssg** [comment]
That cookbook does not look ok. A code snippet shows returning null when in fact you cannot. I was under the impression that you could potentially have multiple ways to authenticate and by returning null the firewall would try another authenticator. That might be a misunderstanding on my part though in which case you're right, it's a documentation problem. 

```php
// src/AppBundle/Security/ApiKeyAuthenticator.php

// ...
use Symfony\Component\Security\Http\HttpUtils;
use Symfony\Component\HttpFoundation\Request;

class ApiKeyAuthenticator implements SimplePreAuthenticatorInterface
{
    protected $userProvider;

    protected $httpUtils;

    public function __construct(UserProviderInterface $userProvider, HttpUtils $httpUtils)
    {
        $this->userProvider = $userProvider;
        $this->httpUtils = $httpUtils;
    }

    public function createToken(Request $request, $providerKey)
    {
        // set the only URL where we should look for auth information
        // and only return the token if we're at that URL
        $targetUrl = '/login/check';
        if (!$this->httpUtils->checkRequestPath($request, $targetUrl)) {
            return;
        }

        // ...
    }
}
```

**Comment by Syam** [comment]
Meet same issue, returning null from CustomAuthenticator::createToken() method will result as a fatal error :

```
Catchable Fatal Error: Argument 1 passed to Symfony\Component\Security\Core\Authentication\AuthenticationProviderManager::authenticate() must be an instance of Symfony\Component\Security\Core\Authentication\Token\TokenInterface, null given, called in (...)/vendor/symfony/symfony/src/Symfony/Component/Security/Http/Firewall/SimplePreAuthenticationListener.php on line 79 (...)
```

**Comment by xelaris** [comment]
@dadamssg I think the bug fixed by #12002 differs from your issue, since there was a `Fatal Error` when passing `null` instead of a token. You get an `AuthenticationCredentialsNotFoundException`, don't you? This is an expected case and should result in a `401 Unauthenticated` response or a redirect to an entry point like a login form. Maybe the missing part is to enable anonymous access to the firewall as shown in http://symfony.com/doc/current/book/security.html#initial-security-yml-setup-authentication?
If so, this indeed should be added as an hint to http://symfony.com/doc/current/cookbook/security/api_key_authentication.html#only-authenticating-for-certain-urls.

> I was under the impression that you could potentially have multiple ways to authenticate and by returning null the firewall would try another authenticator.

As far as I know, it is exaclty as you interpreted, but in your case I guess there is no other authenticator, since your intention is to enable access to urls without any authentication?

To make a long story short: IMO if an `AuthenticationCredentialsNotFoundException` is thrown when returning `null` in your Listener, it is not a bug. I think the doc isn't wrong, since it is possible to return `null`, but apparently there is something missing.

@Syam I think you are describing another issue, since there is a difference in the `AuthenticationCredentialsNotFoundException` described by @dadamssg and the `Fatal Error` you experiencing. Maybe you are using an outdated version which is affected by the issue fixed by #12002 and symfony/symfony@faa8e980?

**Comment by Syam** [comment]
@xelaris This is the same issue as described in this thread title and @dadamssg messages.
I am using symfony 2.5, the fix you referenced is for 2.4

**Comment by xelaris** [comment]
@Syam IMO even if the title matches the same topic, it can't be the same issue, since you get the `Fatal Error`, while he gets an `AuthenticationCredentialsNotFoundException`.
Anyway 2.5 won't get any bugfix, since the support for bugfixes ended in january. Consider upgrading to 2.6.

**Comment by eddiejaoude** [comment]
@xelaris Thanks! Authentication was working but un-authentication was not. 

> Maybe the missing part is to enable anonymous access to the firewall as shown in http://symfony.com/doc/current/book/security.html#initial-security-yml-setup-authentication?

Solved it for me!

**Comment by weaverryan** [comment]
Hi guys!

If you get a "method called on a non-object", then you don't have the fix in #12002 and you'll need to upgrade.

If you're getting the "A Token was not found in the SecurityContext" exception, then you *do* need to add the `anonymous` key to your firewall.

So, I'm going to close this issue, but I've opened up something on the docs: symfony/symfony-docs#5264

Thanks!
